com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationArchiveExtractor.java

```diff
@@ -58,23 +89,66 @@ private void moveStreamToApplicationEntry() throws IOException {
         }
     }
 
-    private void saveEntry(Path filePath) throws IOException {
-        try (OutputStream outputStream = Files.newOutputStream(filePath)) {
-            copy(inputStream, outputStream);
+    private void saveAllEntries(Path dirPath) throws IOException {
+        try (OutputStream fileOutputStream = Files.newOutputStream(dirPath)) {
+            if (isDirectory(moduleFileName)) {
+                saveAsZip(fileOutputStream);
+            } else {
+                saveToFile(fileOutputStream);
+            }
         }
     }
 
-    private void saveAllEntries(Path dirPath) throws IOException {
+    private void saveAsZip(OutputStream fileOutputStream) throws IOException {
+        try (ZipOutputStream zipOutputStream = new ZipOutputStream(fileOutputStream)) {
+            do {
+                if (!hasEntryInCache() && isFile(zipEntryName)) {
+                    zipOutputStream.putNextEntry(createZipEntry());
+                    copy(inputStream, zipOutputStream);
+                    zipOutputStream.closeEntry();
+
+                }
+            } while (getNextEntryByName(moduleFileName) != null);
+        }
+    }
+
+    private ZipEntry createZipEntry() {
+        return new UtcAdjustedZipEntry(getRelativePathOfZipEntry());
+    }
+
+    private String getRelativePathOfZipEntry() {
+        return Paths.get(moduleFileName)
+            .relativize(Paths.get(zipEntryName))
+            .toString();
+    }
+
+    private void saveToFile(OutputStream fileOutputStream) throws IOException {
         do {
-            if (isFile(zipEntryName)) {
-                Path filePath = resolveEntryPath(dirPath);
-                Files.createDirectories(filePath.getParent());
-                filePath = Files.createFile(filePath);
-                saveEntry(filePath);
+            if (!hasEntryInCache()) {
+                copy(inputStream, fileOutputStream);
             }
         } while (getNextEntryByName(moduleFileName) != null);
     }
 
+    private boolean hasEntryInCache() {
+        return knownFileNames.contains(zipEntryName);
+    }
+
+    private String getDigestOfZipEntry(InputStream input, String algorithm) throws DigestException, IOException, NoSuchAlgorithmException {
```

Comment:
Too complicated logic for this class..... Try extract it to another place.

Reply:
done